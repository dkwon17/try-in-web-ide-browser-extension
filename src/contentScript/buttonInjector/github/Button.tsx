/*-----------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Licensed under the MIT License. See LICENSE file in the project root for license information.
 *-----------------------------------------------------------------------------------------------*/

import React from "react";
import { usePopper } from "react-popper";
import { OPEN_OPTIONS } from "../../../messageActions";

import { Endpoint, getActiveEndpoint } from "../../../preferences/preferences";
import { getFactoryURL, getHostName } from "../util";
import { DropdownItem } from "./DropdownItem";
import { DropdownMenu, DropdownMenuDivider } from "./DropdownMenu";
import { EndpointDropdownItem } from "./EndpointDropdownItem";
import "./styles/github.css";

type Props = {
    endpoints: Endpoint[];
    projectURL: string;
    additionalClasses?: string[];
};

/**
 * GitHub button component with dropdown
 */
export const Button = (props: Props) => {
    const [dropdownBtn, _setDropdownBtn] =
        React.useState<HTMLButtonElement | null>(null);
    const [dropdownContent, setDropdownContent] =
        React.useState<HTMLUListElement | null>(null);
    const [showDropdown, setShowDropdown] = React.useState<boolean>(false);

    /**
     * Store dropdown button as reference to use in event listener
     */
    const dropdownBtnRef = React.useRef(dropdownBtn);
    const setDropdownBtn = (element: HTMLButtonElement) => {
        dropdownBtnRef.current = element;
        _setDropdownBtn(element);
    };

    React.useEffect(() => {
        const listener = (event: MouseEvent) => {
            if (event.target instanceof Element) {
                if (dropdownBtnRef.current?.contains(event.target)) {
                    setShowDropdown((prev) => !prev);
                } else {
                    setShowDropdown(false);
                }
            }
        };
        window.addEventListener("click", listener);
        return () => window.removeEventListener("click", listener);
    }, []);

    const { styles, attributes } = usePopper(dropdownBtn, dropdownContent, {
        placement: "bottom-end",
        modifiers: [
            {
                name: "offset",
                options: {
                    offset: [0, 2],
                },
            },
        ],
    });

    const dropdownMenu = showDropdown && (
        <DropdownMenu
            reference={setDropdownContent}
            style={styles.popper}
            attributes={attributes.popper}
        >
            {props.endpoints.map((endpoint: Endpoint) => {
                return (
                    <EndpointDropdownItem
                        endpoint={endpoint}
                        projectURL={props.projectURL}
                        key={endpoint.url}
                    />
                );
            })}

            <DropdownMenuDivider />

            <DropdownItem
                aria-label="Open the extension's options page in a separate tab"
                onClick={() => {
                    chrome.runtime.sendMessage({ action: OPEN_OPTIONS });
                }}
            >
                Configure
            </DropdownItem>
        </DropdownMenu>
    );

    const defaultEndpoint = getActiveEndpoint(props.endpoints);
    const defaultHostname = getHostName(defaultEndpoint);
    const { additionalClasses = [] } = props;

    return (
        <div className={["gh-btn-group"].concat(additionalClasses).join(' ')} id="try-in-web-ide-btn">
            <a
                className="gh-btn gh-btn-padding Button--primary btn-primary Button"
                href={getFactoryURL(props.projectURL, defaultEndpoint)}
                target="_blank"
                aria-label={
                    "Open the current project in a new tab on " +
                    defaultHostname
                }
                title={
                    "Open the current project in a new tab on " +
                    defaultHostname
                }
            >
                Dev Spaces
            </a>
            <button
                type="button"
                ref={setDropdownBtn}
                className="gh-btn btn-primary gh-dropdown-toggle gh-dropdown-toggle-split"
                aria-label="Toggle dropdown"
                title="Toggle dropdown"
            ></button>
            {dropdownMenu}
        </div>
    );
};
