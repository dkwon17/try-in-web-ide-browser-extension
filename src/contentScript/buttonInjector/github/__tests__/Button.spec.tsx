/*-----------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Licensed under the MIT License. See LICENSE file in the project root for license information.
 *-----------------------------------------------------------------------------------------------*/

import { act } from "react-dom/test-utils";
import { render, RenderResult, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";

import { Button } from "../Button";

const projectURL =
    "https://github.com/redhat-developer/try-in-dev-spaces-browser-extension";

describe("Snapshot tests", () => {
    it("renders correctly with one endpoint", () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
        ];

        const { asFragment } = render(
            <Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with multiple endpoints", () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
            { url: "https://url-2.com", active: false, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        const { asFragment } = render(
            <Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with multiple endpoints, with dropdown menu open", async () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
            { url: "https://url-2.com", active: false, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        const { asFragment } = render(
            <Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />
        );
        await openDropdownMenu();

        expect(asFragment()).toMatchSnapshot();
    });
});

describe("Functional tests", () => {

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("should have correct href for main button", async () => {
        let endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        const btn = screen.getByText("Dev Spaces");
        expect((btn as HTMLAnchorElement).href).toEqual(
            "https://url-1.com/f?url=https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        );
    });

    it("should have correct href for main button, multiple endpoints", async () => {
        let endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
            { url: "https://url-2.com", active: false, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        const btn = screen.getByText("Dev Spaces");
        expect((btn as HTMLAnchorElement).href).toEqual(
            "https://url-1.com/f?url=https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        );
    });

    it("should change the href of the button to the active endpoint", async () => {
        let endpoints = [
            { url: "https://url-1.com", active: false, readonly: true },
            { url: "https://url-2.com", active: true, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        const btn = screen.getByText("Dev Spaces");
        expect((btn as HTMLAnchorElement).href).toEqual(
            "https://url-2.com/f?url=https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        );
    });

    it("should display 'Default' label only on the active endpoint", async () => {
        let endpoints = [
            { url: "https://url-1.com", active: false, readonly: true },
            { url: "https://url-2.com", active: true, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        await openDropdownMenu();

        const dropdownMenu = screen.getByTestId("dropdown-menu");
        expect(dropdownMenu.querySelectorAll("a")).toHaveLength(
            endpoints.length + 1
        );
    });

    it("should have the same number of dropdown items as number of endpoints plus one", async () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
            { url: "https://url-2.com", active: false, readonly: false },
            { url: "https://url-3.com", active: false, readonly: false },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        await openDropdownMenu();

        const dropdownMenu = screen.getByTestId("dropdown-menu");
        expect(dropdownMenu.querySelectorAll("a")).toHaveLength(
            endpoints.length + 1
        );
    });

    it("should send message to open options page when configure button is clicked", async () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
        ];

        act(() => {
            render(<Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />);
        });

        await openDropdownMenu();

        const configureBtn = screen.getByText("Configure");
        act(() => {
            configureBtn.click();
        });

        expect(chrome.runtime.sendMessage).toBeCalledWith({
            action: "openOptionsPage",
        });
    });

    it("should remove click event listener on component unmount", async () => {
        const endpoints = [
            { url: "https://url-1.com", active: true, readonly: true },
        ];

        const addListenerSpy = jest.spyOn(window, "addEventListener");
        const removeListenerSpy = jest.spyOn(window, "removeEventListener");

        let renderResult: RenderResult;
        act(() => {
            renderResult = render(
                <Button endpoints={endpoints} projectURL={projectURL} additionalClasses={['ml-2']} />
            );
        });

        expect(addListenerSpy).toBeCalledWith(
            "click",
            expect.any(Function)
        );
        renderResult.unmount();
        expect(removeListenerSpy).toBeCalledWith(
            "click",
            expect.any(Function)
        );
    });
});

async function openDropdownMenu() {
    const dropdownBtn = screen.getByTitle("Toggle dropdown");
    act(() => {
        dropdownBtn.click();
    });

    await waitFor(() => {
        expect(screen.getByTestId("dropdown-menu")).toBeInTheDocument();
    });
}
