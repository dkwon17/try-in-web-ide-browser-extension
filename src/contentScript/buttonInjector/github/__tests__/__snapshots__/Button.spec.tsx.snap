// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests renders correctly with multiple endpoints 1`] = `
<DocumentFragment>
  <div
    class="gh-btn-group ml-2"
    id="try-in-web-ide-btn"
  >
    <a
      class="gh-btn btn-primary"
      href="https://url-1.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
      target="_blank"
      title="Open the project on https://url-1.com"
    >
      Dev Spaces
    </a>
    <button
      class="gh-btn btn-primary gh-dropdown-toggle gh-dropdown-toggle-split"
      title="Toggle dropdown"
      type="button"
    />
  </div>
</DocumentFragment>
`;

exports[`Snapshot tests renders correctly with multiple endpoints, with dropdown menu open 1`] = `
<DocumentFragment>
  <div
    class="gh-btn-group ml-2"
    id="try-in-web-ide-btn"
  >
    <a
      class="gh-btn btn-primary"
      href="https://url-1.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
      target="_blank"
      title="Open the project on https://url-1.com"
    >
      Dev Spaces
    </a>
    <button
      class="gh-btn btn-primary gh-dropdown-toggle gh-dropdown-toggle-split"
      title="Toggle dropdown"
      type="button"
    />
    <ul
      class="gh-dropdown-menu"
      data-popper-escaped="true"
      data-popper-placement="bottom-end"
      data-popper-reference-hidden="true"
      data-testid="dropdown-menu"
      style="position: absolute; left: 0px; top: 0px; right: 0px; transform: translate(0px, 2px);"
    >
      <li
        class="gh-list-item"
      >
        <a
          class="gh-dropdown-item"
          href="https://url-1.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        >
          Open with url-1.com
          <div
            class="gh-pill-badge"
          >
            Default
          </div>
        </a>
      </li>
      <li
        class="gh-list-item"
      >
        <a
          class="gh-dropdown-item"
          href="https://url-2.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        >
          Open with url-2.com
        </a>
      </li>
      <li
        class="gh-list-item"
      >
        <a
          class="gh-dropdown-item"
          href="https://url-3.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
        >
          Open with url-3.com
        </a>
      </li>
      <li>
        <hr
          class="gh-dropdown-divider"
        />
      </li>
      <li
        class="gh-list-item"
      >
        <a
          class="gh-dropdown-item"
        >
          Configure
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Snapshot tests renders correctly with one endpoint 1`] = `
<DocumentFragment>
  <div
    class="gh-btn-group ml-2"
    id="try-in-web-ide-btn"
  >
    <a
      class="gh-btn btn-primary"
      href="https://url-1.com/#https://github.com/redhat-developer/try-in-dev-spaces-browser-extension"
      target="_blank"
      title="Open the project on https://url-1.com"
    >
      Dev Spaces
    </a>
    <button
      class="gh-btn btn-primary gh-dropdown-toggle gh-dropdown-toggle-split"
      title="Toggle dropdown"
      type="button"
    />
  </div>
</DocumentFragment>
`;
